src/python/pants/core/util_rules/archive.py

```diff
@@ -79,31 +98,91 @@ async def find_tar() -> TarBinary:
     return TarBinary(first_path.path, first_path.fingerprint)
 
 
+# -----------------------------------------------------------------------------------------------
+# Create archives
+# -----------------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class CreateArchive:
+    """A request to create an archive.
+
+    All files in the input snapshot will be included in the resulting archive.
+    """
+
+    snapshot: Snapshot
+    output_filename: str
+    format: ArchiveFormat
+
+
+@rule(desc="Creating an archive file", level=LogLevel.DEBUG)
+async def create_archive(
+    request: CreateArchive, tar_binary: TarBinary, zip_binary: ZipBinary
+) -> Digest:
+    if request.format == ArchiveFormat.ZIP:
+        argv = zip_binary.create_archive_argv(request)
+        env = {}
+        input_digest = request.snapshot.digest
+    else:
+        argv = tar_binary.create_archive_argv(request)
+        # `tar` expects to find a couple binaries like `gzip` and `xz` by looking on the PATH.
+        env = {"PATH": os.pathsep.join(SEARCH_PATHS)}
+        # `tar` requires that the output filename's parent directory exists.
```

Comment:
Technically, REAPI guarantees that parent directories of an output filename or directory will exist in the input root. Not sure if the same guarantee is made by Pants local execution. That said, doing this ensures that the invariant holds regardless of what the execution environment does, so seems fine.

Reply:
> Not sure if the same guarantee is made by Pants local execution.

It is.

Reply:
I've had issues in the past with this not being the case, but I can try again to double check.

Reply:
This might not be working properly. When I remove the create dir part from our extraction rule, I get this:

> tar: could not chdir to '__output'